### PR TITLE
[FEATURE,TEST] Adds aligned ends configuration for edit distance alignment.

### PR DESCRIPTION
**Pull request by rrahn**
This enables the aligned ends configuration for the edit distance algorithm.

### CODE REVIEW

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 1]
```diff
@@ -247,8 +247,12 @@ struct alignment_configurator
         auto const & gaps = cfg.template value_or<align_cfg::gap>(gap_scheme{gap_score{-1}});
         auto const & scoring_scheme = get<align_cfg::scoring>(cfg).value;
 
-        // Check if edit distance can be used?
-        if (gaps.get_gap_open_score() == 0)
+        auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+
+        // Don't use edit distance if ...
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: Maybe negate the documentation

```
        // Only use edit distance if ...
        if (gaps.get_gap_open_score() == 0 &&  // gap open score is not set.
            !(align_ends_cfg[2] || align_ends_cfg[3]) && // free end gaps are not set for second sequence.
            align_ends_cfg[0] == align_ends_cfg[1]) // free ends for leading and trailing gaps do not differ in first sequence
```

I think it's easier to follow

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp [hunk 2]
```diff
@@ -80,6 +59,7 @@ struct default_edit_distance_trait_type
 {
     //!\brief The default word type.
     using word_type = uint64_t;
+    using is_semi_global_type = std::false_type;
```
Comment (eseiler) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: This is not documented

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 3]
```diff
@@ -282,10 +286,61 @@ struct alignment_configurator
...
+        // Configure semi-global alignment
+        // ----------------------------------------------------------------------------
+
+        // Get the value for the sequence ends configuration.
+        auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+        using align_ends_cfg_t = remove_cvref_t<decltype(align_ends_cfg)>;
+
+        auto configure_with_semi_global = [&](auto is_semi_global)
+        {
+
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: remove newline

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 4]
```diff
@@ -247,8 +247,12 @@ struct alignment_configurator
...
-        // Check if edit distance can be used?
-        if (gaps.get_gap_open_score() == 0)
+        auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+
+        // Only use edit distance if ...
+        if (gaps.get_gap_open_score() == 0 &&  // gap open score is not set,
+            !(align_ends_cfg[2] || align_ends_cfg[3]) && // none of the free end gaps are set for second seq,
+            align_ends_cfg[0] == align_ends_cfg[1]) // free ends for leading and trailing gaps are equal in first seq.
         {
             // TODO: Instead of relying on nucleotide scoring schemes we need to be able to determine the edit distance
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: Is this in a card or issue?

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 5]
```diff
@@ -282,10 +286,61 @@ struct alignment_configurator
...
+        // Get the value for the sequence ends configuration.
+        auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+        using align_ends_cfg_t = remove_cvref_t<decltype(align_ends_cfg)>;
+
+        auto configure_with_semi_global = [&](auto is_semi_global)
+        {
+
+            struct edit_traits_type
+            {
+                using word_type           [[maybe_unused]] = uint_fast64_t;
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: Do I need maybe unused in struct member variables? :o

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp [hunk 6]
```diff
@@ -71,6 +46,10 @@ template <typename traits_type>
 SEQAN3_CONCEPT edit_distance_trait_concept = requires
 {
     typename std::remove_reference_t<traits_type>::word_type;
+    typename std::remove_reference_t<traits_type>::is_semi_global_type;
+
+    // Must be a boolean integral constant.
+    requires std::is_same_v<typename std::remove_reference_t<traits_type>::is_semi_global_type::value_type, bool>;
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: `std::Same`?

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp [hunk 7]
```diff
@@ -139,15 +120,14 @@ class pairwise_alignment_edit_distance_unbanded
...
+    static constexpr bool is_semi_global = traits_t::is_semi_global_type::value;
     //!\brief Whether the alignment is a global alignment or not.
-    static constexpr bool is_global = detail::global_config_concept<align_config_t> && !is_semi_global;
+    static constexpr bool is_global = !is_semi_global;
 
     //!\brief How to pre-initialize hp.
     static constexpr word_type hp0 = is_global ? 1 : 0;
 
     static_assert(8 * sizeof(word_type) <= 64, "we assume at most uint64_t as word_type");
-    static_assert((is_global && !is_semi_global) || (!is_global && is_semi_global), "Either set global or semi-global");
```
Comment (smehringer) [inline include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: why did you remove this? Can I do both at the same time?

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp (hunk 5 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp]: Well inside of the local trait the compiler (only gcc-7 btw) would emit an unused variable warning

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp (hunk 4 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp]: yep

File: include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp (hunk 7 above)
Comment (rrahn) [reply include/seqan3/alignment/pairwise/edit_distance_unbanded.hpp]: it is not possible to have set both at the same time with the new logic so the assertion is always true and therefore redundant.

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 8]
```diff
@@ -282,10 +286,42 @@ struct alignment_configurator
...
-            throw invalid_alignment_configuration{"Free end gaps is not yet supported."};
+        // ----------------------------------------------------------------------------
+        // Configure semi-global alignment
+        // ----------------------------------------------------------------------------
+
+        // Get the value for the sequence ends configuration.
+        auto align_ends_cfg = cfg.template value_or<align_cfg::aligned_ends>(align_cfg::none_ends_free);
+        using align_ends_cfg_t = remove_cvref_t<decltype(align_ends_cfg)>;
+
+        auto configure_semi_global = [&](auto is_semi_global)
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: I would name this differently

Maybe

```suggestion
        auto configure_edit_traits = [&](auto is_semi_global)
```

File: include/seqan3/alignment/pairwise/alignment_configurator.hpp [hunk 9]
```diff
@@ -282,10 +286,42 @@ struct alignment_configurator
...
+            else // Resolve correct property at runtime.
+                return configure_semi_global(align_ends_cfg[1] ? std::true_type{} : std::false_type{});
+        };
+
+        // Check if it has free ends set for the first sequence leading gaps.
+        // If possible use static information.
+        if constexpr (align_ends_cfg_t::template is_static<0>())
+            return has_free_ends_trailing(std::integral_constant<bool, align_ends_cfg_t::template get_static<0>()>{});
+        else // Resolve correct property at runtime.
+            return has_free_ends_trailing(align_ends_cfg[0] ? std::true_type{} : std::false_type{});
```
Comment (marehr) [inline include/seqan3/alignment/pairwise/alignment_configurator.hpp]: I doubt that this works :)

```suggestion
            if (align_ends_cfg[0])
                return has_free_ends_trailing(std::true_type{});
            else
                return has_free_ends_trailing(std::false_type{});
```

Comment (rrahn) [reply include/seqan3/alignment/pairwise/alignment_configurator.hpp]: Me too 🤣
